streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java

```diff
@@ -0,0 +1,230 @@
+     * @param <K>         key type
+     * @return The {@link StreamsMetadata} for the storeName and key
+     */
+    public synchronized <K> StreamsMetadata getMetadataWithKey(final String storeName,
+                                                               final K key,
+                                                               final StreamPartitioner<K, ?> partitioner) {
+        if (storeName == null) {
+            throw new IllegalArgumentException("storeName cannot be null");
+        }
+
+        if (key == null) {
+            throw new IllegalArgumentException("key cannot be null");
+        }
+
+        if (partitioner == null) {
+            throw new IllegalArgumentException("partitioner cannot be null");
+        }
+
+        final Set<String> sourceTopics = builder.stateStoreNameToSourceTopics().get(storeName);
+        if (sourceTopics == null) {
+            return null;
+        }
+
+
+        int numPartitions = 0;
+        for (String topic : sourceTopics) {
+            final List<TopicPartition> topicPartitions = partitionsByTopic.get(topic);
+            if (topicPartitions.size() > numPartitions) {
+                numPartitions = topicPartitions.size();
+            }
+        }
+
+        final Integer partition = partitioner.partition(key, null, numPartitions);
+        final Set<TopicPartition> matchingPartitions = new HashSet<>();
+        for (String sourceTopic : sourceTopics) {
+            matchingPartitions.add(new TopicPartition(sourceTopic, partition));
+        }
+
+        for (StreamsMetadata streamsMetadata : allMetadata) {
+            final Set<String> stateStoreNames = streamsMetadata.stateStoreNames();
+            final Set<TopicPartition> topicPartitions = new HashSet<>(streamsMetadata.topicPartitions());
+            topicPartitions.retainAll(matchingPartitions);
+            if (stateStoreNames.contains(storeName)
+                    && !topicPartitions.isEmpty()) {
+                return streamsMetadata;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Respond to changes to the HostInfo -> TopicPartition mapping. Will rebuild the
+     * metadata
+     * @param currentState  the current mapping of {@link HostInfo} -> {@link TopicPartition}s
+     */
+    public synchronized void onChange(final Map<HostInfo, Set<TopicPartition>> currentState) {
+        rebuildPartitionsByTopic(currentState);
+        rebuildInstances(currentState);
+    }
+
+    private boolean hasPartitionsForAnyTopics(final Set<String> topicNames, final Set<TopicPartition> partitions) {
+        for (String topic : topicNames) {
+            for (TopicPartition tp : partitionsByTopic.get(topic)) {
+                if (partitions.contains(tp)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+
+    private void rebuildPartitionsByTopic(final Map<HostInfo, Set<TopicPartition>> currentState) {
+        partitionsByTopic.clear();
```

Comment:
Could we get rid of this `partitionsByTopic` map, and in `hasPartitionsForAnyTopics` just check if `TopicPartition.topic().equals(topic)`?